javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -12,245 +12,455 @@
+  override def add(e: E): Boolean =
+    RB.insert(tree, e, ()) == null
+
+  override def remove(o: Any): Boolean =
+    RB.delete(tree, o) != null
+
+  override def clear(): Unit =
+    RB.clear(tree)
+
+  def subSet(fromElement: E, fromInclusive: Boolean, toElement: E,
+      toInclusive: Boolean): NavigableSet[E] = {
+    new Projection(tree,
+        fromElement, RB.boundKindFromIsInclusive(fromInclusive),
+        toElement, RB.boundKindFromIsInclusive(toInclusive))
+  }
+
+  def headSet(toElement: E, inclusive: Boolean): NavigableSet[E] = {
+    new Projection(tree,
+        null.asInstanceOf[E], RB.NoBound,
+        toElement, RB.boundKindFromIsInclusive(inclusive))
   }
 
-  private val inner: mutable.TreeSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+  def tailSet(fromElement: E, inclusive: Boolean): NavigableSet[E] = {
+    new Projection(tree,
+        fromElement, RB.boundKindFromIsInclusive(inclusive),
+        null.asInstanceOf[E], RB.NoBound)
+  }
 
-  def iterator(): Iterator[E] = {
-    new Iterator[E] {
-      private val iter = inner.clone.iterator
+  def subSet(fromElement: E, toElement: E): SortedSet[E] =
+    subSet(fromElement, true, toElement, false)
 
-      private var last: Option[E] = None
+  def headSet(toElement: E): SortedSet[E] =
+    headSet(toElement, false)
 
-      def hasNext(): Boolean = iter.hasNext
+  def tailSet(fromElement: E): SortedSet[E] =
+    tailSet(fromElement, true)
 
-      def next(): E = {
-        last = Some(iter.next().inner)
-        last.get
-      }
+  def comparator(): Comparator[_ >: E] =
+    NaturalComparator.unselect(comp)
 
-      def remove(): Unit = {
-        if (last.isEmpty) {
-          throw new IllegalStateException()
-        } else {
-          last.foreach(self.remove(_))
-          last = None
-        }
-      }
+  def first(): E = {
+    if (isEmpty())
+      throw new NoSuchElementException()
+    RB.minKey(tree)
+  }
+
+  def last(): E = {
+    if (isEmpty())
+      throw new NoSuchElementException()
+    RB.maxKey(tree)
+  }
+
+  def lower(e: E): E =
+    RB.maxKeyBefore(tree, e, RB.ExclusiveBound)
+
+  def floor(e: E): E =
+    RB.maxKeyBefore(tree, e, RB.InclusiveBound)
+
+  def ceiling(e: E): E =
+    RB.minKeyAfter(tree, e, RB.InclusiveBound)
+
+  def higher(e: E): E =
+    RB.minKeyAfter(tree, e, RB.ExclusiveBound)
+
+  def pollFirst(): E = {
+    val node = RB.minNode(tree)
+    if (node ne null) {
+      RB.deleteNode(tree, node)
+      node.key
+    } else {
+      null.asInstanceOf[E]
+    }
+  }
+
+  def pollLast(): E = {
+    val node = RB.maxNode(tree)
+    if (node ne null) {
+      RB.deleteNode(tree, node)
+      node.key
+    } else {
+      null.asInstanceOf[E]
     }
   }
 
-  def descendingIterator(): Iterator[E] = {
-    new Iterator[E] {
-      private val iter = inner.iterator.toList.reverse.iterator
+  override def clone(): TreeSet[E] =
+    new TreeSet(tree.treeCopy())(comp)
+}
+
+private object TreeSet {
+  private final class Projection[E](
```

Comment:
Have you considered sharing the parts of the order independent implementation? (some of the naming might need to change from `upper/lower` to `to/from`).